src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncPivotBlockSelector.scala

```diff
@@ -0,0 +1,167 @@
+
+    val peersSortedByBestNumber = peersUsedToChooseTarget.toList.sortBy { case (_, number) => -number }
+    val bestPeerBestBlockNumber = peersSortedByBestNumber.headOption
+      .map { case (_, bestPeerBestBlockNumber) => bestPeerBestBlockNumber }
+      .getOrElse(BigInt(0))
+    val expectedPivotBlock = (bestPeerBestBlockNumber - syncConfig.pivotBlockOffset).max(0)
+    val peersToAsk = peersSortedByBestNumber
+      .takeWhile { case (_, number) => number >= expectedPivotBlock }
+      .map { case (peer, _) => peer }
+
+    if (peersToAsk.size >= minPeersToChoosePivotBlock) {
+      log.info(
+        "Trying to choose fast sync pivot block using {} peers. The best block is {}. Ask {} peers for block nr {}",
+        peersUsedToChooseTarget.size,
+        bestPeerBestBlockNumber,
+        peersToAsk.size,
+        expectedPivotBlock
+      )
+
+      peersToAsk.foreach { peer =>
+        peerEventBus ! Subscribe(MessageClassifier(Set(BlockHeaders.code), PeerSelector.WithId(peer.id)))
+        etcPeerManager ! EtcPeerManagerActor.SendMessage(
+          GetBlockHeaders(Left(expectedPivotBlock), 1, 0, reverse = false),
+          peer.id
+        )
+      }
+
+      val timeout = scheduler.scheduleOnce(peerResponseTimeout, self, PivotBlockTimeout)
+      context become waitingForPivotBlock(peersToAsk.map(_.id).toSet, expectedPivotBlock, timeout, Map.empty)
+    } else {
+      log.info(
+        "Cannot pick pivot block. Need at least {} peers, but there are only {} which meet the criteria ({} all available at the moment). Retrying in {}",
+        minPeersToChoosePivotBlock,
+        peersToAsk.size,
+        peersUsedToChooseTarget.size,
+        startRetryInterval
+      )
+      scheduleRetry(startRetryInterval)
+    }
+  }
+
+  def waitingForPivotBlock(
+      peersToAsk: Set[PeerId],
+      targetBlockNumber: BigInt,
+      timeout: Cancellable,
+      headers: Map[ByteString, BlockHeaderWithVotes]
+  ): Receive =
+    handleCommonMessages orElse {
+      case MessageFromPeer(blockHeaders: BlockHeaders, peerId) =>
+        peerEventBus ! Unsubscribe(MessageClassifier(Set(BlockHeaders.code), PeerSelector.WithId(peerId)))
+        val updatedPeersToAsk = peersToAsk - peerId
+        val targetBlockHeaderOpt =
+          if (blockHeaders.headers.size != 1) None
+          else
+            blockHeaders.headers.find(header => header.number == targetBlockNumber)
+        targetBlockHeaderOpt match {
+          case Some(targetBlockHeader) =>
+            log.info("Received vote for {} from {}", targetBlockHeader.hashAsHexString, peerId.value)
+            val newValue =
+              headers.get(targetBlockHeader.hash).map(_.vote).getOrElse(BlockHeaderWithVotes(targetBlockHeader))
+            val updatedHeaders = headers.updated(targetBlockHeader.hash, newValue)
+            val BlockHeaderWithVotes(mostPopularBlockHeader, updatedVotes) = updatedHeaders.mostVotedHeader
+            if (updatedVotes >= minPeersToChoosePivotBlock) {
```

Comment:
If i understand correctly how this works now is that even if we have lets say 7 peers to whom we send request, we finish selection when we receive `minPeersToChoosePivotBlock` same blocks, and we do not wait for the rest of responses ?
I wonder if this does not make such scenario possible:
1.  We do our selection and receive enough votes for block, but we do not receive header from lets say `peer1`
2. We are back to fast sync and start making requests, and we make request for x block headers from `peer1`, it means we create `PeerRequestHandler` which subsribes to `PeerEventBus` for `BlockHeaders` messages from `Peer1` .
3. `peer1` respond with pivot block. `PivotBlockSelector` do not receive response as it already finished, but `PeerRequestHandler` from fast sync headers request is still alive and receievs this pivot block
4. we blacklist `peer1` for response which do not match our fast sync request.

Reply:
I think it is possible. But it is the same case when we assume the timeout - this response could be received later. I could add collecting for all msgs if you prefer, but in a case when we have a lot of peers it won't perform very well